### PR TITLE
Use statistics in Faker CTAS

### PR DESCRIPTION
**Pull request by nineinchnick**
## Description

Use statistics when using `CREATE TABLE AS SELECT` in the Faker connector to:
* set the `default_limit` table property to the estimated number of rows from the source table
* create a view to read from the table with range predicates, based on the statistics
* detect high-cardinality `BIGINT` columns and use sequences for them
* detect low-cardinality columns and generate dictionaries to select values from

## Release notes

( ) This is not user-visible or is docs only, and no release notes are required.
(x) Release notes are required. Please propose a release note for me.
( ) Release notes are required, with the following suggested text:

```markdown
## Faker
* Use statistics when using `CREATE TABLE AS SELECT` in the Faker connector. ({issue}`issuenumber`)
```

### CODE REVIEW

File: plugin/trino-faker/src/main/java/io/trino/plugin/faker/FakerMetadata.java [hunk 1]
```diff
@@ -28,76 +33,112 @@
...
 import static java.util.Objects.requireNonNull;
+import static java.util.stream.Collectors.joining;
+import static java.util.stream.Collectors.toMap;
 
 public class FakerMetadata
         implements ConnectorMetadata
 {
     public static final String SCHEMA_NAME = "default";
     public static final String ROW_ID_COLUMN_NAME = "$row_id";
+    private static final String VIEW_SUFFIX = "_view";
```
Comment (nineinchnick) [inline plugin/trino-faker/src/main/java/io/trino/plugin/faker/FakerMetadata.java:112]: This should probably be configurable, but I need suggestions for config option name, default value, etc.

File: plugin/trino-faker/src/main/java/io/trino/plugin/faker/FakerMetadata.java [hunk 2]
```diff
@@ -371,12 +440,304 @@ public synchronized Optional<ConnectorOutputMetadata> finishCreateTable(Connecto
...
+        List<String> viewColumns = columns.stream()
+                .filter(column -> !ROW_ID_COLUMN_NAME.equals(column.name()))
+                .map(column -> {
+                    String name = column.name();
+                    Type type = types.get(name);
+                    if (!type.equals(BIGINT)) {
+                        return quoted(name);
+                    }
+                    // distinct values is an approximation
+                    if ((double) distinctValues.get(name) / rowCount.get() < 0.98) {
```
Comment (nineinchnick) [inline plugin/trino-faker/src/main/java/io/trino/plugin/faker/FakerMetadata.java:504]: The threshold might be configurable

File: plugin/trino-faker/src/main/java/io/trino/plugin/faker/FakerMetadata.java [hunk 3]
```diff
@@ -371,12 +440,304 @@ public synchronized Optional<ConnectorOutputMetadata> finishCreateTable(Connecto
...
+                                                true)),
+                                        false)))),
+                0,
+                maxDictionarySize)) {
+            // TODO support reading values from multiple pages
+            Page page = null;
+            while (page == null) {
+                page = pageSource.getNextPage();
+            }
+            checkState(maxDictionarySize <= page.getPositionCount(), "Max dictionary size cannot be greater than %d".formatted(page.getPositionCount()));
```
Comment (nineinchnick) [inline plugin/trino-faker/src/main/java/io/trino/plugin/faker/FakerMetadata.java:613]: Should I add a `@Max()` in the configuration and schema/table property validators?

File: plugin/trino-faker/src/main/java/io/trino/plugin/faker/FakerMetadata.java [hunk 4]
```diff
@@ -297,7 +327,7 @@ public synchronized FakerOutputTableHandle beginCreateTable(ConnectorSession ses
         double schemaNullProbability = (double) schema.properties().getOrDefault(SchemaInfo.NULL_PROBABILITY_PROPERTY, nullProbability);
         double tableNullProbability = (double) tableMetadata.getProperties().getOrDefault(TableInfo.NULL_PROBABILITY_PROPERTY, schemaNullProbability);
 
-        ImmutableList.Builder<ColumnInfo> columns = ImmutableList.builder();
+        ImmutableList.Builder<ColumnInfo> columnsBuilder = ImmutableList.builder();
```
Comment (losipiuk) [inline plugin/trino-faker/src/main/java/io/trino/plugin/faker/FakerMetadata.java:364]: revert?

File: plugin/trino-faker/src/main/java/io/trino/plugin/faker/FakerMetadata.java [hunk 5]
```diff
@@ -371,12 +406,110 @@ public synchronized Optional<ConnectorOutputMetadata> finishCreateTable(Connecto
...
+    public synchronized void createView(
+            ConnectorSession session,
+            SchemaTableName viewName,
+            ConnectorViewDefinition definition,
+            Map<String, Object> viewProperties,
+            boolean replace)
+    {
+        checkArgument(viewProperties.isEmpty(), "This connector does not support creating views with properties");
+        checkSchemaExists(viewName.getSchemaName());
+        if (views.containsKey(viewName) && !replace) {
```
Comment (losipiuk) [inline plugin/trino-faker/src/main/java/io/trino/plugin/faker/FakerMetadata.java:673]: you need also to check if such table exists

File: plugin/trino-faker/src/main/java/io/trino/plugin/faker/FakerMetadata.java [hunk 6]
```diff
@@ -371,12 +406,110 @@ public synchronized Optional<ConnectorOutputMetadata> finishCreateTable(Connecto
...
+    @Override
+    public synchronized void setViewColumnComment(ConnectorSession session, SchemaTableName viewName, String columnName, Optional<String> comment)
+    {
+        ConnectorViewDefinition view = getView(session, viewName).orElseThrow(() -> new ViewNotFoundException(viewName));
+        views.put(viewName, new ConnectorViewDefinition(
+                view.getOriginalSql(),
+                view.getCatalog(),
+                view.getSchema(),
+                view.getColumns().stream()
+                        .map(currentViewColumn -> Objects.equals(columnName, currentViewColumn.getName()) ? new ConnectorViewDefinition.ViewColumn(currentViewColumn.getName(), currentViewColumn.getType(), comment) : currentViewColumn)
```
Comment (losipiuk) [inline plugin/trino-faker/src/main/java/io/trino/plugin/faker/FakerMetadata.java:709]: nit: use `columnName.equals` not `Objects.equals`

File: plugin/trino-faker/src/main/java/io/trino/plugin/faker/FakerMetadata.java [hunk 7]
```diff
@@ -407,10 +424,99 @@ public synchronized Optional<ConnectorOutputMetadata> finishCreateTable(
         requireNonNull(info, "info is null");
 
         tables.put(tableName, new TableInfo(info.columns(), info.properties(), info.comment()));
+        createViewWithPredicates(tableName, info.columns(), computedStatistics);
 
         return Optional.empty();
     }
 
+    private synchronized void createViewWithPredicates(SchemaTableName tableName, List<ColumnInfo> columns, Collection<ComputedStatistics> computedStatistics)
```
Comment (losipiuk) [inline plugin/trino-faker/src/main/java/io/trino/plugin/faker/FakerMetadata.java]: can you call out to standard view creation logic to reuse already-exists checks?

File: plugin/trino-faker/src/main/java/io/trino/plugin/faker/FakerMetadata.java (hunk 1 above)
Comment (nineinchnick) [reply plugin/trino-faker/src/main/java/io/trino/plugin/faker/FakerMetadata.java:112]: Yes, I think that would be too implicit. Maybe if we'd use some procedure instead of `CTAS`. The view is also supposed to be an example or a starting point - it can be recreated with additional predicates, better projections, etc.

File: plugin/trino-faker/src/main/java/io/trino/plugin/faker/FakerMetadata.java [hunk 8]
```diff
@@ -457,13 +467,30 @@ else if (metadata.getStatisticType().equals(ColumnStatisticType.MAX_VALUE)) {
...
-        String viewSql = "SELECT * FROM %s.%s %s".formatted(
+        List<String> viewColumns = columns.stream()
+                .filter(column -> !ROW_ID_COLUMN_NAME.equals(column.name()))
+                .map(column -> {
+                    String name = column.name();
+                    Type type = types.get(name);
+                    if (!type.equals(BIGINT) || !minimums.containsKey(name)) {
+                        return quoted(name);
+                    }
+                    if ((double) distinctValues.get(name) / rowCount.get() < 1) {
```
Comment (losipiuk) [inline plugin/trino-faker/src/main/java/io/trino/plugin/faker/FakerMetadata.java]: I do not get why are you using `< 1` here.
Should it be rather something like `abs(1.0 - (double) distinctValues.get(name) / rowCount.get()) < some_epsilon`?

File: plugin/trino-faker/src/main/java/io/trino/plugin/faker/FakerMetadata.java [hunk 9]
```diff
@@ -425,6 +425,15 @@ public synchronized Optional<ConnectorOutputMetadata> finishCreateTable(
         TableInfo info = tables.get(tableName);
         requireNonNull(info, "info is null");
 
+        if (!computedStatistics.isEmpty() && !info.properties().containsKey(TableInfo.DEFAULT_LIMIT_PROPERTY)) {
+            ComputedStatistics statistic = computedStatistics.stream().findFirst().orElseThrow();
+            long rowCount = (long) getValue(BIGINT, 0, statistic.getTableStatistics().get(TableStatisticType.ROW_COUNT));
```
Comment (losipiuk) [inline plugin/trino-faker/src/main/java/io/trino/plugin/faker/FakerMetadata.java:445]: nit: static import `ROW_COUNT`

File: plugin/trino-faker/src/main/java/io/trino/plugin/faker/FakerMetadata.java [hunk 10]
```diff
@@ -425,6 +425,15 @@ public synchronized Optional<ConnectorOutputMetadata> finishCreateTable(
...
+            info = info.withProperties(ImmutableMap.<String, Object>builder()
+                    .putAll(info.properties())
+                    .put(TableInfo.DEFAULT_LIMIT_PROPERTY, rowCount)
```
Comment (losipiuk) [inline plugin/trino-faker/src/main/java/io/trino/plugin/faker/FakerMetadata.java:448]: nit: static import `DEFAULT_LIMIT_PROPERTY`

File: plugin/trino-faker/src/main/java/io/trino/plugin/faker/FakerConnector.java [hunk 11]
```diff
@@ -136,6 +144,14 @@ public List<PropertyMetadata<?>> getTableProperties()
...
                         null,
                         defaultLimit -> checkProperty(1 <= defaultLimit, INVALID_TABLE_PROPERTY, "default_limit value must be equal or greater than 1"),
+                        false),
+                longProperty(
+                        TableInfo.MAX_DICTIONARY_SIZE,
+                        """
+                                Maximum size of randomly generated dictionaries to pick values from, used for columns with low number of approximate distinct values
+                                observed during table creation using existing data. Set to zero to disable using dictionaries""",
+                        null,
+                        maxDictionarySize -> checkProperty(0 <= maxDictionarySize, INVALID_TABLE_PROPERTY, "max_dictionary_size value must be equal or greater than 0"),
```
Comment (losipiuk) [inline plugin/trino-faker/src/main/java/io/trino/plugin/faker/FakerConnector.java:154]: This property looks weird on table level. Should you rather have per-column property saying if the column is dictionary or not?

**Review by losipiuk — COMMENTED** [review]
Nice!

Added some comments. Feel free to ignore most

File: plugin/trino-faker/src/main/java/io/trino/plugin/faker/FakerConnector.java (hunk 11 above)
Comment (nineinchnick) [reply plugin/trino-faker/src/main/java/io/trino/plugin/faker/FakerConnector.java:154]: This allows disabling dictionaries for an individual table. Column properties cannot be set when running a CTAS.

File: plugin/trino-faker/src/main/java/io/trino/plugin/faker/FakerMetadata.java (hunk 8 above)
Comment (nineinchnick) [reply plugin/trino-faker/src/main/java/io/trino/plugin/faker/FakerMetadata.java]: Must have been late when I wrote this, this shows I need better test cases.

File: plugin/trino-faker/src/main/java/io/trino/plugin/faker/FakerPageSource.java [hunk 12]
```diff
@@ -38,8 +38,10 @@
 import io.trino.spi.type.TimestampType;
 import io.trino.spi.type.TimestampWithTimeZoneType;
 import io.trino.spi.type.Type;
+import io.trino.spi.type.UuidType;
```
Comment (raunaqmorarka) [inline plugin/trino-faker/src/main/java/io/trino/plugin/faker/FakerPageSource.java:41]: Maybe make a separate PR for this commit ? This looks okay to merge already

File: plugin/trino-faker/src/main/java/io/trino/plugin/faker/FakerMetadata.java (hunk 10 above)
Comment (nineinchnick) [reply plugin/trino-faker/src/main/java/io/trino/plugin/faker/FakerMetadata.java:448]: This will conflict with the schema property.

File: plugin/trino-faker/src/main/java/io/trino/plugin/faker/FakerPageSource.java (hunk 12 above)
Comment (nineinchnick) [reply plugin/trino-faker/src/main/java/io/trino/plugin/faker/FakerPageSource.java:41]: https://github.com/trinodb/trino/pull/24146

File: plugin/trino-faker/src/test/java/io/trino/plugin/faker/TestFakerMetadata.java [hunk 13]
```diff
@@ -0,0 +1,346 @@
...
+import static io.trino.spi.security.PrincipalType.USER;
+import static io.trino.spi.type.BigintType.BIGINT;
+import static io.trino.testing.QueryAssertions.assertEqualsIgnoreOrder;
+import static io.trino.testing.TestingConnectorSession.SESSION;
+import static io.trino.testing.assertions.TrinoExceptionAssert.assertTrinoExceptionThrownBy;
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.assertThatThrownBy;
+import static org.assertj.core.api.Fail.fail;
+
+public class TestFakerMetadata
```
[the remark this answers is not in the document]
Comment (nineinchnick) [reply plugin/trino-faker/src/test/java/io/trino/plugin/faker/TestFakerMetadata.java:47]: I copied these tests from the memory connector, since I also copied most of the view management implementation. Can we refactor the tests in a follow-up PR for both connectors?

File: plugin/trino-faker/src/main/java/io/trino/plugin/faker/FakerMetadata.java (hunk 7 above)
Comment (nineinchnick) [reply plugin/trino-faker/src/main/java/io/trino/plugin/faker/FakerMetadata.java]: Yes, thanks!

File: plugin/trino-faker/src/main/java/io/trino/plugin/faker/FakerMetadata.java (hunk 8 above)
Comment (nineinchnick) [reply plugin/trino-faker/src/main/java/io/trino/plugin/faker/FakerMetadata.java]: This is actually correct - I just changed 1 to 0.98, I should definitely make it a constant, but I was wondering if it also should be configurable.

File: plugin/trino-faker/src/main/java/io/trino/plugin/faker/FakerMetadata.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (nineinchnick) [reply plugin/trino-faker/src/main/java/io/trino/plugin/faker/FakerMetadata.java]: I used a view to avoid parsing the predicate values from strings. I actually like how we can generate the view and allow users to fine-tune it, without adding a bunch of separate column properties, that would be as hard to edit as the view.

File: plugin/trino-faker/src/main/java/io/trino/plugin/faker/FakerMetadata.java (hunk 3 above)
Comment (losipiuk) [reply plugin/trino-faker/src/main/java/io/trino/plugin/faker/FakerMetadata.java:605]: If you can express upper bound as a constant - sure.

File: plugin/trino-faker/src/main/java/io/trino/plugin/faker/FakerMetadata.java [hunk 14]
```diff
@@ -371,12 +450,284 @@ public synchronized Optional<ConnectorOutputMetadata> finishCreateTable(Connecto
...
+                        return quoted(name);
+                    }
+                    // distinct values is an approximation
+                    if ((double) distinctValues.get(name) / rowCount.get() <= tableMinSequenceRatio) {
+                        return quoted(name);
+                    }
+                    if (!minimums.containsKey(name)) {
+                        return "\"$row_id\" AS " + quoted(name);
+                    }
+                    return "\"$row_id\" + " + formatLiteral(type, minimums.get(name)) + " AS " + quoted(name);
```
Comment (losipiuk) [inline plugin/trino-faker/src/main/java/io/trino/plugin/faker/FakerMetadata.java:519]: do you want to handle maximums too? (spread the values somehow?)

**Comment by nineinchnick** [comment]
#24147 describes that we should not be using predicate pushdown for constraining data generation, so I'll rework this using table/column properties only.
